bulkio: Make incremental scheduled backup wait for full backup.

miretskiy opened the pull request:
Fixes #52835

Add ability to record schedule groups: set of related schedules.
Use this functionality to makean incremental schedule wait
until the full one completes before it begins its execution.

Release Notes: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52836)
<!-- Reviewable:end -->

miretskiy commented:
@dt -- rebased on master; ready for review

dt commented on pkg/ccl/backupccl/backup_job.go:
```diff
@@ -571,24 +572,45 @@ func (b *backupResumer) Resume(
+			ctx,
+			"lookup-schedule-info",
+			txn,
+			sqlbase.InternalExecutorSessionDataOverride{User: security.RootUser},
```
this should probably be `NodeUser`

dt commented on pkg/ccl/backupccl/create_scheduled_backup.go:
```diff
@@ -320,54 +307,69 @@ func doCreateBackupSchedules(
+			if err != nil {
+				return err
+			}
+			inc.Pause("Waiting for FULL")
```
"Waiting for initial backup to complete"

dt reviewed (commented):
Hm. I donno, this might just be me losing my marbles after staring at too much code, so feel free to just say so, but on first read, I'm about dubious of introducing new concepts like "ScheduleGroup" and "GroupAction".  Again, might be just me, but I find the it is a bit harder to grok with the new concepts. Or, alternatively, I expect them to be actual rows in a table called `schedule_groups` and then to have a `group_id` column in schedules or whatever. I also don't know that `GroupAction` is really _general_ to all schedule types?

Thus, my counterproposal would be: 
1) the _backup specific_ `ScheduledBackupExecutionArgs` gains a "optional int64 unpause_on_success` field 
2) the success hook for backups looks at it and, if it is set: a) unpauses the schedule with that ID b) clears it from itself.

WDYT?

miretskiy reviewed (commented):
Sure -- that makes sense.  I had something like this going -- notify success is a bit tricky w/ jobs; and initially, I was having issues w/ doing it in the resumer and kinda got on a path of doing it through "meta" proto... Eventually, backup resumer on success started working fine - so making these changes backup specific is fine by me.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52836#-:-MF50YqOCUMAIPkDSUMe:bes34qj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)

___
*[pkg/ccl/backupccl/backup_job.go, line 590 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52836#-MEiYs0iyOK0U3ZEuLm5-r1-590:-MF52gFI4ddytWV80sGm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbe70eeffc206304bdccda013f53b2f9bbbc738b/pkg/ccl/backupccl/backup_job.go#L590)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

this should probably be `NodeUser`
</blockquote></details>

Done.
___
*[pkg/ccl/backupccl/create_scheduled_backup.go, line 341 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52836#-MEiYs0iyOK0U3ZEuLm6-r1-341:-MF54vmEEW7ageh63pRO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbe70eeffc206304bdccda013f53b2f9bbbc738b/pkg/ccl/backupccl/create_scheduled_backup.go#L341)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

"Waiting for initial backup to complete"
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dt reviewed (approved):
Looks like you picked up the go 1.14 proto diff

miretskiy commented:
Argh... Will fix.

On Wed, Aug 19, 2020 at 7:15 AM David Taylor <notifications@github.com>
wrote:

> *@dt* approved this pull request.
>
> Looks like you picked up the go 1.14 proto diff
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/52836#pullrequestreview-470357273>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANA4FVDDSR4GQED3QSSU5C3SBOX55ANCNFSM4P72C6AA>
> .
>

Commit:
bulkio: Make incremental scheduled backup wait for full backup.

Add ability to record schedule groups: set of related schedules.
Use this functionality to makean incremental schedule wait
until the full one completes before it begins its execution.

Release Notes: None

miretskiy commented:
bors r+
